[7주차] 보안

tbnsok40 opened the pull request:
7주차 열심히 달려보겠습니다. 질문 많이 남길 수 있는 한 주로 만들겠습니다!

**로그인이 필요없는 API**
로그인 - POST /session
회원 생성하기 - POST /users
고양이 장난감 목록 얻기 - GET /products
고양이 장난감 상세 조회하기 - GET /products/{id}

**로그인이 필요한 API**
고양이 장난감 등록하기 - POST /products
고양이 장난감 수정하기 - PATCH /products/{id}
고양이 장난감 삭제하기 - DELETE /products/{id}
회원 수정하기 - POST /users/{id}
회원 삭제하기 - DELETE /users/{id}

**요구 사항**
테스트 커버리지 100%를 달성해야 합니다.
모든 API 테스트를 통과해야 합니다.
모든 E2E 테스트를 통과해야 합니다.

tbnsok40 committed:
Spring Security 7-1강 수강

tbnsok40 commented on app/src/main/java/com/codesoom/assignment/config/SecurityJavaConfig.java:
```diff
@@ -0,0 +1,28 @@
+        Filter authenticationFilter = new JwtAuthenticationFilter(authenticationManager(), authenticationService);
+        Filter authenticationErrorFilter = new AuthenticationErrorFilter();
+        http.csrf().disable()
+                .addFilter(authenticationFilter)
+                .addFilterBefore(authenticationErrorFilter,
+                        JwtAuthenticationFilter.class);
```
addFilterBefore()메소드는, 메소드의 두번째 인자가 실행되기 전에 첫번째 인자가 실행되도록 동작한다고 배웠습니다.
두번째 인자는 결국 인스턴스로 authenticationFilter 를 생성하여 addFilter()에 인자로 들어가는데, 그럼 굳이 addFilter() 메소드가 필요한 이유가 무엇인지 궁금합니다..!

hannut91 replied:
저도 대충 알고 있었는데 궁금해서 찾아봤습니다.

> Allows adding a Filter before one of the known Filter classes. The known Filter instances are either a Filter listed in HttpSecurityBuilder.addFilter(Filter) or a Filter that has already been added using HttpSecurityBuilder.addFilterAfter(Filter, Class) or HttpSecurityBuilder.addFilterBefore(Filter, Class).

즉 `addFilter`로 추가되어있는 filter앞에 추가하는 것이므로, `addFilter`를 해야지만 추가되어있는 것들중에 찾아서 앞에 추가할 수 있습니다.

See also
  - https://docs.spring.io/spring-security/site/docs/current/api/org/springframework/security/config/annotation/web/builders/HttpSecurity.html#addFilterBefore(javax.servlet.Filter,java.lang.Class)

hannut91 commented on app/src/main/java/com/codesoom/assignment/filters/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,74 @@
+        String authorization = request.getHeader("Authorization");
+        String accessToken;
+
+        try {
+            if (authorization == null) {
+                throw new InvalidTokenException(""); // 빈 토큰을 넣어준다 어차피 예외경우니
+            }
+            accessToken = authorization.substring("Bearer ".length());
+            authenticationService.parseToken(accessToken);
+
+        } catch (InvalidTokenException exception) {
+            response.sendError(HttpStatus.UNAUTHORIZED.value());
+            return;
+        }
```
여기서 `catch`를 하지 않으면 어떻게 되나요?

tbnsok40 replied:
캐치 구문이 없다면, 유효하지 않은 토큰으로 접근 시 예외처리를 던지지 못할 것 같습니다.

tbnsok40 replied:
catch 구문을 제거해도 예외처리가 되도록, SecurityJavaConfig.java 내부에 addFilter(authenticationErrorFilter) 를 해주었습니다.  authenticationErrorFilter 에서 InvalidToken의 예외 처리를 담당한다고 이해하였습니다.

tbnsok40 committed:
SessionCreationPolicy STATLESS 지정, ExceptionHandling() 추가

tbnsok40 committed:
`@RequestAttributes` 어노테이션 삭제

tbnsok40 committed:
JwtAuthenticationFilter, try~catch 구문 삭제

tbnsok40 committed:
UserAuthentication클래스 생성

tbnsok40 committed:
오타 수정

tbnsok40 commented on app/src/main/java/com/codesoom/assignment/config/SecurityJavaConfig.java:
```diff
@@ -23,6 +29,15 @@ protected void configure(HttpSecurity http) throws Exception{
                 .addFilterBefore(authenticationErrorFilter,
-                        JwtAuthenticationFilter.class);
+                        JwtAuthenticationFilter.class)
+                .sessionManagement()
+                    .sessionCreationPolicy(SessionCreationPolicy.STATELESS)
```
SessionCreationPolicy 의 네가지 Policy 를 찾아보았습니다.

- SessionCreationPolicy.Always   :  스프링 시큐리티가 항상 세션 생성
- SessionCreationPolicy.If_Required   :  스프링 시큐리티가 필요 시 세션 생성(기본값)
- SessionCreationPolicy.Never     :  스프링 시큐리티가 세션을 생성하지 않지만, 이미 존재하면 사용
- `SessionCreationPolicy.Stateless`    :  스프링 시큐리티가 세션을 생성하지 않고, 존재해도 사용하지 않음 -> 그렇다면 다른 곳에서 세션을 생성 -> JWT 토큰방식 사용하는 경우에 해당

tbnsok40 replied:
Stateless 는 쿠키가 사용되지 않고, 매 요청마다 재인증을 해야합니다. 민감한 로그인 정보이기 때문에 쿠키를 저장하지 않고 매번 재요청(재 로그인)해야 하는 것이 맞다고 이해하였습니다.
또한 이러한 무상태성(stateless)이 REST API의 stateless 제약조건에 부합한다고 공식문서에서 찾을 수 있었습니다.

tbnsok40 committed:
UPDATE, DELETE `@PreAuthorize()` 추가

tbnsok40 committed:
PasswordEncoder bean 등록

tbnsok40 committed:
PasswordEncoder 생성자 주입

tbnsok40 committed:
changePassword() 메서드 명 encodePassword() 로 변경

hannut91 commented:
Spring Security를 이용해서 잘 구현해주셨습니다. 완벽하게 이해하지 않아도 구현할 수 있지만, 내부적으로 어떻게 동작하는지 이해한다면 더 잘 구현할 수 있겠죠. Spring Security에 대해서 더 살펴보시면 좋을 것 같습니다. 다만 Spring Security를 잘한다고 해서 스프링을 잘하는건 아니니, 너무 깊게보시는 건 권장하진 않습니다.

tbnsok40 committed:
UserServiceTest BDD 적용

tbnsok40 committed:
given() 구문 재배치

tbnsok40 commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -64,105 +73,160 @@ void setUp() {
+
+                assertThat(user.getId()).isEqualTo(13L);
+                assertThat(user.getEmail()).isEqualTo("tester@example.com");
+                assertThat(user.getName()).isEqualTo(NAME_BEFORE_UPDATE);
+                assertThat(user.getPassword()).isEqualTo("");
```
트레이너님 안녕하세요 :0) 
98번째 줄에서 비밀번호는 어떻게 출력되는지 확인하고자, 
isEqualTo() 내부에 일부러 틀리게 문자를 넣고 테스트를 하였습니다.

확인 결과 아예 비어있다고 출력되어 이 점이 조금 의아했습니다.  혹시 제가 어떤 부분을 생각해보면 좋은지, 어느 점을 놓쳤는지 피드백을 받고싶습니다..!

tbnsok40 committed:
UserNotMatchException 클래스 생성
본인이 작성한 글만 수정/삭제 할 수 있도록 예외 클래스를 작성합니다.

tbnsok40 committed:
UserControllerTest updateUserWithValidAttributes() 메소드 테스트 실패
에러메시지로 ` nested exception is java.lang.NullPointerException: Cannot invoke "com.codesoom.assignment.domain.User.getId()" because "user" is null`가 발생하였습니다.
어느 부분에서 부터 꼬였는지 이해하지 못해 어떻게 질문해야 할지를 잘 모르겠습니다 죄송합니다..
user가 존재하지 않기 때문에 getId() 를 하지 못한다 까지 이해했는데, 왜 user가 null인지 잘 모르겠습니다..

tbnsok40 commented on app/src/main/java/com/codesoom/assignment/controllers/UserController.java:
```diff
@@ -28,11 +30,15 @@ UserResultData create(@RequestBody @Valid UserRegistrationData registrationData)
             @PathVariable Long id,
-            @RequestBody @Valid UserModificationData modificationData
+            @RequestBody @Valid UserModificationData modificationData,
+            UserAuthentication authentication
```
커밋 메세지 내용대로 user가 왜 Null 인지 찾기 위해 처음부터 다시 코드를 작성해보았습니다.

update 메소드에 UserAuthentication 인자를 추가한 후에, 모든 update() 메소드에 세번째 인자로 userId를 추가하면서부터 NullPointerException 에러가 발생하는걸 알게됐습니다.

그렇다면 UserAuthentication 클래스에서 찾고자하는 User를 getUserId() 메소드로 못찾는 것이라 의심되는데, 그 원인에 대해 질문드리고 싶습니다..!

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -103,7 +118,9 @@ void updateUserWithValidAttributes() throws Exception {
                         .contentType(MediaType.APPLICATION_JSON)
-                        .content("{\"name\":\"TEST\",\"password\":\"test\"}")
+                        .content("{\"name\":\"TEST\",\"password\":\"TEST\"}")
+                        .header("Authorization", "Bearer" + MY_TOKEN)
```
```suggestion
                        .header("Authorization", "Bearer " + MY_TOKEN)
```

여기에 띄어쓰기가 잘못되어있었던 것 같네요

hannut91 replied:
그렇다고 하더라도 왜 NullPointerException이 발생했고, 왜 테스트에서 실패했는지 쉽게 찾을 수 없었는지 고민해 볼 필요가 있습니다.

코드를 하나씩 따라가봤는데, 일단 토큰이 잘못되어서 토큰을 처리할 때 잘못되는 것 같습니다.

```java
    public Long parseToken(String accessToken) {
        Claims claims = jwtUtil.decode(accessToken);
        return claims.get("userId", Long.class);
    }
```

저는 일단 여기서 토큰이 잘못되었을 경우 로직이 끝날 줄 알았는데, 계속 되더라구요. 그리고 사용자의 ID는 `0`으로 되어있었습니다.

그래서 Mocking 한 부분을 보니

```java
        given(authenticationService.parseToken(MY_TOKEN)).willReturn(1L);
        given(authenticationService.parseToken(OTHER_TOKEN)).willReturn(2L);
```

위와 같이 되어있는데, `MY_TOKEN`에 앞에 빈 칸 하나가 들어가니, Long의 기본값 0L이 반환되어서 그랬던 거였네요

tbnsok40 committed:
UserController의 update, delete 권한 추가

hannut91 commented:
> 그래도 조금이라도 하고 자자며 이불을 박차고 일어났다

좋은 액션입니다 ㅎㅎ 저는 이 때의 감정을 잊지 않는게 중요하다고 생각합니다.

빈 칸 하나 때문에 고생이 많으셨군요! 테스트에서 가독성만큼 중요한게 무엇이 잘못되었을 때 얼마나 쉽게 알아차릴 수 있는가도 매우 중요합니다. 저도 그래서 테스트가 깨졌을 때 왜 빨리 알아채지 못했지?로 접근했던 것 같습니다.
